### ISSUE TITLE
Invalid memory access when trying to access HID devices

### ISSUE THREAD

**Issue by mrwedders** [qzind/tray#119]
Hi everyone,

I've been using QZ Tray (2.0.x, built from source a couple of weeks ago) for printing without any issues, but my employer recently got a bunch of Dymo USB scales which didn't have any bundled software to send the reading to any application. After going round the houses a bit with WinUSB in C I realised QZ can do this too!

My implementation is working fine, except it occasionally fails to list HID devices (qz.hid.listDevices) giving the Java error...

```
2016-10-17 15:01:17,716 [ERROR] Connection error
java.lang.RuntimeException: Cannot call method public void qz.ws.PrintSocketClient#onMessage(org.eclipse.jetty.websocket.api.Session, java.io.Reader) with args: [org.eclipse.jetty.websocket.common.WebSocketSession, org.eclipse.jetty.websocket.common.message.MessageReader]
    at org.eclipse.jetty.websocket.common.events.annotated.CallableMethod.unwrapRuntimeException(CallableMethod.java:93)
    at org.eclipse.jetty.websocket.common.events.annotated.CallableMethod.call(CallableMethod.java:75)
    at org.eclipse.jetty.websocket.common.events.annotated.OptionalSessionCallableMethod.call(OptionalSessionCallableMethod.java:68)
    at org.eclipse.jetty.websocket.common.events.JettyAnnotatedEventDriver$2.run(JettyAnnotatedEventDriver.java:210)
    at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
    at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
    at java.lang.Thread.run(Unknown Source)
Caused by: java.lang.Error: Invalid memory access
    at com.sun.jna.Native.invokeVoid(Native Method)
    at com.sun.jna.Function.invoke(Function.java:374)
    at com.sun.jna.Function.invoke(Function.java:323)
    at com.sun.jna.Library$Handler.invoke(Library.java:236)
    at com.sun.proxy.$Proxy6.hid_free_enumeration(Unknown Source)
    at org.hid4java.jna.HidApi.freeEnumeration(HidApi.java:179)
    at org.hid4java.HidDeviceManager.getAttachedHidDevices(HidDeviceManager.java:249)
    at org.hid4java.HidServices.getAttachedHidDevices(HidServices.java:167)
    at qz.communication.HidUtilities.getHidDevices(Unknown Source)
    at qz.communication.HidUtilities.getHidDevicesJSON(Unknown Source)
    at qz.ws.PrintSocketClient.processMessage(Unknown Source)
    at qz.ws.PrintSocketClient.onMessage(Unknown Source)
    at sun.reflect.GeneratedMethodAccessor10.invoke(Unknown Source)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
    at java.lang.reflect.Method.invoke(Unknown Source)
    at org.eclipse.jetty.websocket.common.events.annotated.CallableMethod.call(CallableMethod.java:70)
    ... 5 more
```

When I reproduced this error QZ Tray needed to be restarted before it would successfully list HID devices again.

Any ideas?

**Comment by tresf** [comment]
@mrwedders which OS?  So far we've had reports of this on Windows XP but none others.

I also assume you're on either 2.0.1 or the latest compiled from 2.0 branch.  I'll quote the client we're working on this bug with... 


> Just on the surface of the error message “invalid memory access” does sound like a broken pointer to me. This line :
> > Balance a `hid_device_io_start()` call with `hid_device_io_stop()` in the
error path. This avoids processing of HID reports when the probe fails
which possibly leads to invalid memory access in `hid_device_probe()` as
`report_enum->report_id_hash` might already be freed via
`hid_close_report()`.

> From http://code.metager.de/source/history/linux/stable/drivers/hid/
> [...]

**Comment by mrwedders** [comment]
This is on Windows 7 x64, AMD A6-3500, 8Gb RAM, running Java 8u101. I reproduced it on another Windows 7 x64 machine with similar hardware. I've now got the error logs from the PC where the issue first appeared and the errors are identical (well, $Proxy7 rather than $Proxy6, but I'm guessing that's immaterial).

This is a compile from the 2.0 branch, 2-3 weeks old now. I'll rebuild from the latest 2.0 and 2.1 sources tonight and will see if they make any difference tomorrow.

The only other thing I can think of is this machine is not up to date with Windows Update since it only accesses our internal system, could that be causing a problem? I think last updates were installed some time in 2014.

**Comment by tresf** [comment]
> could that be causing a problem? I think last updates were installed some time in 2014.

I don't know the answer to this question.  The HID library is a C++ library written by another team.  We can escalate issues with community support or issue pull requests against their code if the problem is specific to the upstream library.

On that note, if you can reproduce, are there code snippets you can provide?  In the case of our other report of this same exact issue it was very intermittent and there wasn't a way to reliably reproduce. 

The client is using HID and printing at the same time and constantly chattering back and forth.  We've [patched the websocket side to help](https://github.com/qzind/tray/commit/a6e71e92dcd46fef40d6e34110a7e3a240e7e4b6), but you already have that fix so there still may be room for improvement.

**Comment by mrwedders** [comment]
I managed to reproduce it by basically repeating the process that gets the weight until it happened - it didn't take that many attempts! I'll have a better play with it tomorrow and see if I can come up with a reliable way to reproduce it.

I could also try letting one of the machines run Windows Update to see if it makes any difference, but that might take some time!

All the code is basically adapted from the QZ sample page. I call qz.hid.listDevices every second until I find the right device, then claim it and open a stream. The window beforeunload event closes the stream, releases the device and disconnects from QZ Tray when the user navigates away. I have attached this but I don't think there's much going on in there (
[qz_dymo_js.txt](https://github.com/qzind/tray/files/534018/qz_dymo_js.txt)).

I'm not trying to do any printing at this stage - that occurs at the next page of our dispatch process, the only common Javascript is the qz.security methods.

**Comment by tresf** [comment]
@mrwedders have you considered using [`startListening`](https://qz.io/api/qz.hid#.startListening) if the device isn't found rather than the looping method?

There are a few potential race conditions I see:

1. When navigating away from a page `beforeunload` you may still have a `setTimeout(self.qzFindDymoUSBScale, 1000);` lingering.  You should find a way to cancel that so a scheduled poll doesn't come in.  That should almost never happen but you should code against that scenario.
2. You aren't promisifying the `closeStream.then(releaseDevice(...))`, so you may have a race condition there as well.
3. You are calling `releaseDevice` even if the device isn't claimed.  You should check [`isClaimed`](https://qz.io/api/qz.hid#.isClaimed) or else you may throw an error and never close the websocket.

**Comment by mrwedders** [comment]
I built the latest 2.0.x and 2.1 sources last night, and I've messed around with this a lot on my Mac... 

I'm now closing the stream, releasing the device and disconnecting from QZ Tray as soon as I've got a valid weight, and QZ Tray is still crashing regularly when it tries to claim the device even with this completely clean exit. It's not even logging an error, the last line in the logs is "[INFO] Allowed LR Printing Root to claim a USB device".
It seems to crash mostly when the device has been disconnected - which is an issue as the scales I'm using power off automatically after a few minutes of inactivity!

I've also tried startListening but that seems very unreliable. As soon as I make a claimDevice or listDevices call I don't get any more attach/detach events, even if I stop and restart the listening. I have to do one of those things to see if the device is connected when the page loads, so I don't see a way around that problem.

As I say this is all on Mac, I haven't tested any of this on a PC but hopefully will today. My updated JS is here: [qz_dymo_js.txt](https://github.com/qzind/tray/files/538307/qz_dymo_js.txt)

**Comment by mrwedders** [comment]
I have just spent a while doing repeated measures on the Windows 7 PCs that were initially experiencing crashes with the new code posted above and haven't been able to cause a crash - possibly this is a Mac only issue? 
I might set up one of the staff with the new scales later on and see if they have any problems during the day.

**Comment by tresf** [comment]
> possibly this is a Mac only issue? 

Perhaps.  Is is the same exception thrown or something new?  https://github.com/gary-rowe/hid4java/issues/37

**Comment by mrwedders** [comment]
As I said, it isn't logging an exception in the debug.log files, the last line just says it's allowed me to claim the USB device.

Here ([qz_crash_mac_19-10-16.txt](https://github.com/qzind/tray/files/539217/qz_crash_mac_19-10-16.txt)) is the output from OSX's "QZ Tray quit unexpectedly" dialog. This is with most recent Javascript given above, current 2.0.x branch, and I can cause a crash very quickly by turning the device on and off a few times. This is currently being tested by one user on a Windows 7 machine and it seems to be functioning perfectly.

If there's a better way of getting debug to you let me know - I'm mostly a Windows guy so I'm not that familiar with debugging crashes on a Mac!

**Comment by tresf** [comment]
> If there's a better way of getting debug to you let me know 

The preferred method is to run from command line.  This varies by platform.
https://qz.io/wiki/faq#debugging-qz-tray

Generally the crash will reference a `hs_err_pid` file, which an be pasted to gist or uploaded as an attachment.  If this crash shows to be a duplicate of https://github.com/qzind/tray/pull/25#issuecomment-211726043 then we will want to track it as a separate issue and close this one out as resolved.

**Comment by tresf** [comment]
@mrwedders perhaps we're discussing two separate issues here... 

> `Caused by: java.lang.Error: Invalid memory access`

> This is on Windows 7 x64, AMD A6-3500, 8Gb RAM, running Java 8u101. I reproduced it on another Windows 7 x64 machine with similar hardware. I've now got the error logs from the PC where the issue first appeared and the errors are identical (well, $Proxy7 rather than $Proxy6, but I'm guessing that's immaterial).

We have another client experiencing this on Windows XP and Windows Vista.  

We've recently obtained a Dymo USB scale and the device seems to power itself off quite often.  Releasing and reclaiming the device seems to help, but this seems like a problem perpetuated by sleepy hardware.

We will continue to investigate, but any help with this -- such as exact steps to reproduce, insight into the upstream bug -- is appreciated.

**Comment by tresf** [comment]
@mrwedders a quick summary:

 * Mac crashes are inevitable.  They won't go away until we sort something with either Apple or a workaround with signal11, neither of which have any movement on, sorry!
 * PC crashes are being worked on.  We have a [tree using CodeMinders API that can be tested right now](https://github.com/qzind/tray/tree/hid3) as well as as [tree using PureJavaHIDAPI](https://github.com/tresf/tray/tree/hid) which can be tested right now.

Since these crashes are intermittent (we cannot reproduce in our lab) we need much more testing in order to resolve this problem.  Please let us know what your testing reveals.

**Comment by tresf** [comment]
@mrwedders the original bug (Windows) will be fixed with the switch to a new HID library: `nyholku/purejavahidapi`.  If you want to test it out now, you can compile it here: https://github.com/tresf/tray/tree/hid

If you'd rather not compile, this fix will be part of the 2.0.3 release.

Note, Mac and Linux will continue to use `gary-rowe/hid4java` in 2.0.3.  Only Windows will be switching to `nyholku/purejavahidapi`.

**Comment by tresf** [comment]
Closed via https://github.com/qzind/tray/commit/b6da1a8fc3d96946c6a0e642c176c4dd99768100.
